src/infra/core/src/auth/authentication_service_impl.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use core::panic;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use chrono::{Duration, Utc};
+use dill::component;
+use internal_error::{ErrorIntoInternal, InternalError};
+use jsonwebtoken::errors::ErrorKind;
+use jsonwebtoken::{decode, encode, Algorithm, DecodingKey, EncodingKey, Header, Validation};
+use kamu_core::auth::*;
+use rand::distributions::Alphanumeric;
+use rand::{thread_rng, Rng};
+use serde::{Deserialize, Serialize};
+
+///////////////////////////////////////////////////////////////////////////////
+
+const KAMU_JWT_ISSUER: &str = "dev.kamu";
+const KAMU_JWT_ALGORITHM: Algorithm = Algorithm::HS384;
+
+///////////////////////////////////////////////////////////////////////////////
+
+pub struct AuthenticationServiceImpl {
+    kamu_jwt_secret: String,
+    authentication_providers_by_method: HashMap<&'static str, Arc<dyn AuthenticationProvider>>,
+}
+
+#[component(pub)]
+impl AuthenticationServiceImpl {
+    pub fn new(authentication_providers: Vec<Arc<dyn AuthenticationProvider>>) -> Self {
+        let kamu_jwt_secret = match std::env::var("KAMU_JWT_SECRET") {
+            Ok(jwt_secret) => jwt_secret,
+            Err(_) => Self::random_jwt_secret(),
```

Comment:
I'm usually scared of such fallbacks - one small typo and our backend is silently miss-configured.

Should we make it a required DI argument?

Reply:
Having a coordinated JWT secret would be necessary for scaling API server instances.
It could be randomized at Kubernetes level, however, a different secret would mean dropping existing sessions after restart.
On the other hand, secret has to be periodially randomized, so some rotation rules are needed.

Reply:
An open question is also whether to externalize the issuer and signing algorithm properties.